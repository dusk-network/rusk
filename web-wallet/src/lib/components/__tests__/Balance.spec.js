import { afterEach, describe, expect, it } from "vitest";
import { cleanup, render } from "@testing-library/svelte";
import { skipIn } from "lamb";
import { Balance } from "..";

describe("Balance", () => {
  const baseProps = {
    fiatCurrency: "USD",
    fiatPrice: 10,
    locale: "en",
    publicBalance: 2_000_000n,
    shieldedBalance: 1_000_000n,
    tokenCurrency: "DUSK",
  };

  const baseOptions = {
    props: baseProps,
    target: document.body,
  };

  afterEach(cleanup);

  it("renders the Balance component", () => {
    const { container } = render(Balance, baseOptions);

    expect(
      container.querySelector(".dusk-balance__usage-details")
    ).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should update the Balance component when the props change", async () => {
    const { container, rerender } = render(Balance, baseOptions);

    expect(container.firstChild).toMatchSnapshot();

    await rerender({
      fiatCurrency: "EUR",
      fiatPrice: 20,
      locale: "it",
      publicBalance: 2_500_000n,
      shieldedBalance: 500_000n,
      tokenCurrency: "DUSK",
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should display the right percentage values", async () => {
    const { container } = render(Balance, baseOptions);

    // Check if shielded percentage displays as 33.33%
    expect(
      container.querySelector(
        ".dusk-balance__account:first-child .dusk-balance__percentage"
      )?.textContent
    ).toContain("33.33%");

    // Check if public percentage displays as 66.67%
    expect(
      container.querySelector(
        ".dusk-balance__account:last-child .dusk-balance__percentage"
      )?.textContent
    ).toContain("66.67%");
  });

  it("should display the right percentage values when balance is zero", async () => {
    const options = {
      ...baseOptions,
      props: { ...baseProps, publicBalance: 0n, shieldedBalance: 0n },
    };

    const { container } = render(Balance, options);

    expect(
      container.querySelector(
        ".dusk-balance__account:first-child .dusk-balance__percentage"
      )?.textContent
    ).toContain("0%");

    expect(
      container.querySelector(
        ".dusk-balance__account:last-child .dusk-balance__percentage"
      )?.textContent
    ).toContain("0%");
  });

  it("should pass additional class names and attributes to the rendered element", async () => {
    const props = {
      ...baseProps,
      className: "foo bar",
      id: "balance",
    };
    const { container, rerender } = render(Balance, { ...baseOptions, props });

    expect(container.firstChild).toHaveClass("foo bar");
    expect(container.firstChild).toHaveAttribute("id", "balance");

    await rerender({
      ...props,
      className: "qux",
      id: "new-balance",
    });

    expect(container.firstChild).toHaveClass("qux");
    expect(container.firstChild).toHaveAttribute("id", "new-balance");
  });

  it("should not display the fiat value if the fiat price is `undefined`", () => {
    const props = skipIn(baseProps, ["fiatPrice"]);
    const { container } = render(Balance, { ...baseOptions, props });

    expect(container.querySelector(".dusk-balance__fiat--visible")).toBeNull();
    expect(container.firstChild).toMatchSnapshot();
  });
});
