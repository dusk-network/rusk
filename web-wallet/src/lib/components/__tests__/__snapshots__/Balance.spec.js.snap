// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Balance > renders the Balance component 1`] = `
<article
  class="dusk-balance"
>
  <header
    class="dusk-balance__header"
  >
    <h2
      class="sr-only"
    >
      Your Balance:
    </h2>
  </header>
   
  <p
    class="dusk-balance__dusk"
  >
    <span>
      0.003000000
    </span>
     
    <span>
      DUSK
    </span>
  </p>
   
  <p
    class="dusk-balance__fiat dusk-balance__fiat--visible"
  >
    <span>
      $0.03
    </span>
  </p>
   
  <div
    class="dusk-card dusk-card--gap-default dusk-balance__usage-details"
  >
     
    <div
      class="dusk-card__body-container"
    >
      <div
        class="dusk-balance__account"
      >
        <span
          class="dusk-balance__percentage"
        >
          <svg
            class="dusk-icon dusk-icon--size--default"
            data-tooltip-id="main-tooltip"
            data-tooltip-text="Shielded"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M12,1L3,5V11C3,16.55 6.84,21.74 12,23C17.16,21.74 21,16.55 21,11V5L12,1M12,7C13.4,7 14.8,8.1 14.8,9.5V11C15.4,11 16,11.6 16,12.3V15.8C16,16.4 15.4,17 14.7,17H9.2C8.6,17 8,16.4 8,15.7V12.2C8,11.6 8.6,11 9.2,11V9.5C9.2,8.1 10.6,7 12,7M12,8.2C11.2,8.2 10.5,8.7 10.5,9.5V11H13.5V9.5C13.5,8.7 12.8,8.2 12,8.2Z"
            />
          </svg>
          
          33.33
          %
        </span>
         
        <span
          class="dusk-balance__value"
        >
          0.001000000
          <svg
            class="dusk-icon dusk-icon--size--default"
            data-tooltip-id="main-tooltip"
            data-tooltip-text="DUSK"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M12.335 0.005a12.186 12.186 0 0 0 -2.58 0.204C4.2 1.259 0 6.14 0 11.999 0 17.859 4.203 22.74 9.759 23.79c0.726 0.138 1.476 0.21 2.241 0.21 6.789 0 12.261 -5.64 11.991 -12.49C23.741 5.249 18.599 0.173 12.335 0.005Zm0.207 21.58c-0.204 0.012 -0.336 -0.222 -0.21 -0.384 1.974 -2.541 3.153 -5.733 3.153 -9.202 0 -3.468 -1.176 -6.663 -3.153 -9.202 -0.126 -0.162 0.003 -0.396 0.207 -0.384 5.052 0.282 9.06 4.465 9.06 9.586 0 5.121 -4.005 9.304 -9.057 9.586Z"
            />
          </svg>
          
        </span>
      </div>
       
      <div
        class="dusk-balance__account"
      >
        <span
          class="dusk-balance__percentage"
        >
          <svg
            class="dusk-icon dusk-icon--size--default"
            data-tooltip-id="main-tooltip"
            data-tooltip-text="Public"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M21 11C21 16.5 17.2 21.7 12 23C6.8 21.7 3 16.5 3 11V5L12 1L21 5V11M12 21C15.8 20 19 15.5 19 11.2V6.3L12 3.2L5 6.3V11.2C5 15.5 8.2 20 12 21M14.8 11H10.5V8.5C10.5 7.7 11.2 7.2 12 7.2S13.5 7.7 13.5 8.5V9H14.8V8.5C14.8 7.1 13.4 6 12 6S9.2 7.1 9.2 8.5V11C8.6 11 8 11.6 8 12.2V15.7C8 16.4 8.6 17 9.2 17H14.7C15.4 17 16 16.4 16 15.8V12.3C16 11.6 15.4 11 14.8 11Z"
            />
          </svg>
          
          66.67
          %
        </span>
         
        <span
          class="dusk-balance__value"
        >
          0.002000000
          <svg
            class="dusk-icon dusk-icon--size--default"
            data-tooltip-id="main-tooltip"
            data-tooltip-text="DUSK"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M12.335 0.005a12.186 12.186 0 0 0 -2.58 0.204C4.2 1.259 0 6.14 0 11.999 0 17.859 4.203 22.74 9.759 23.79c0.726 0.138 1.476 0.21 2.241 0.21 6.789 0 12.261 -5.64 11.991 -12.49C23.741 5.249 18.599 0.173 12.335 0.005Zm0.207 21.58c-0.204 0.012 -0.336 -0.222 -0.21 -0.384 1.974 -2.541 3.153 -5.733 3.153 -9.202 0 -3.468 -1.176 -6.663 -3.153 -9.202 -0.126 -0.162 0.003 -0.396 0.207 -0.384 5.052 0.282 9.06 4.465 9.06 9.586 0 5.121 -4.005 9.304 -9.057 9.586Z"
            />
          </svg>
          
        </span>
      </div>
    </div>
     
  </div>
  
</article>
`;

exports[`Balance > should not display the fiat value if the fiat price is \`undefined\` 1`] = `
<article
  class="dusk-balance"
>
  <header
    class="dusk-balance__header"
  >
    <h2
      class="sr-only"
    >
      Your Balance:
    </h2>
  </header>
   
  <p
    class="dusk-balance__dusk"
  >
    <span>
      0.003000000
    </span>
     
    <span>
      DUSK
    </span>
  </p>
   
  <p
    class="dusk-balance__fiat"
  >
    <span>
      $0.00
    </span>
  </p>
   
  <div
    class="dusk-card dusk-card--gap-default dusk-balance__usage-details"
  >
     
    <div
      class="dusk-card__body-container"
    >
      <div
        class="dusk-balance__account"
      >
        <span
          class="dusk-balance__percentage"
        >
          <svg
            class="dusk-icon dusk-icon--size--default"
            data-tooltip-id="main-tooltip"
            data-tooltip-text="Shielded"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M12,1L3,5V11C3,16.55 6.84,21.74 12,23C17.16,21.74 21,16.55 21,11V5L12,1M12,7C13.4,7 14.8,8.1 14.8,9.5V11C15.4,11 16,11.6 16,12.3V15.8C16,16.4 15.4,17 14.7,17H9.2C8.6,17 8,16.4 8,15.7V12.2C8,11.6 8.6,11 9.2,11V9.5C9.2,8.1 10.6,7 12,7M12,8.2C11.2,8.2 10.5,8.7 10.5,9.5V11H13.5V9.5C13.5,8.7 12.8,8.2 12,8.2Z"
            />
          </svg>
          
          33.33
          %
        </span>
         
        <span
          class="dusk-balance__value"
        >
          0.001000000
          <svg
            class="dusk-icon dusk-icon--size--default"
            data-tooltip-id="main-tooltip"
            data-tooltip-text="DUSK"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M12.335 0.005a12.186 12.186 0 0 0 -2.58 0.204C4.2 1.259 0 6.14 0 11.999 0 17.859 4.203 22.74 9.759 23.79c0.726 0.138 1.476 0.21 2.241 0.21 6.789 0 12.261 -5.64 11.991 -12.49C23.741 5.249 18.599 0.173 12.335 0.005Zm0.207 21.58c-0.204 0.012 -0.336 -0.222 -0.21 -0.384 1.974 -2.541 3.153 -5.733 3.153 -9.202 0 -3.468 -1.176 -6.663 -3.153 -9.202 -0.126 -0.162 0.003 -0.396 0.207 -0.384 5.052 0.282 9.06 4.465 9.06 9.586 0 5.121 -4.005 9.304 -9.057 9.586Z"
            />
          </svg>
          
        </span>
      </div>
       
      <div
        class="dusk-balance__account"
      >
        <span
          class="dusk-balance__percentage"
        >
          <svg
            class="dusk-icon dusk-icon--size--default"
            data-tooltip-id="main-tooltip"
            data-tooltip-text="Public"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M21 11C21 16.5 17.2 21.7 12 23C6.8 21.7 3 16.5 3 11V5L12 1L21 5V11M12 21C15.8 20 19 15.5 19 11.2V6.3L12 3.2L5 6.3V11.2C5 15.5 8.2 20 12 21M14.8 11H10.5V8.5C10.5 7.7 11.2 7.2 12 7.2S13.5 7.7 13.5 8.5V9H14.8V8.5C14.8 7.1 13.4 6 12 6S9.2 7.1 9.2 8.5V11C8.6 11 8 11.6 8 12.2V15.7C8 16.4 8.6 17 9.2 17H14.7C15.4 17 16 16.4 16 15.8V12.3C16 11.6 15.4 11 14.8 11Z"
            />
          </svg>
          
          66.67
          %
        </span>
         
        <span
          class="dusk-balance__value"
        >
          0.002000000
          <svg
            class="dusk-icon dusk-icon--size--default"
            data-tooltip-id="main-tooltip"
            data-tooltip-text="DUSK"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M12.335 0.005a12.186 12.186 0 0 0 -2.58 0.204C4.2 1.259 0 6.14 0 11.999 0 17.859 4.203 22.74 9.759 23.79c0.726 0.138 1.476 0.21 2.241 0.21 6.789 0 12.261 -5.64 11.991 -12.49C23.741 5.249 18.599 0.173 12.335 0.005Zm0.207 21.58c-0.204 0.012 -0.336 -0.222 -0.21 -0.384 1.974 -2.541 3.153 -5.733 3.153 -9.202 0 -3.468 -1.176 -6.663 -3.153 -9.202 -0.126 -0.162 0.003 -0.396 0.207 -0.384 5.052 0.282 9.06 4.465 9.06 9.586 0 5.121 -4.005 9.304 -9.057 9.586Z"
            />
          </svg>
          
        </span>
      </div>
    </div>
     
  </div>
  
</article>
`;

exports[`Balance > should update the Balance component when the props change 1`] = `
<article
  class="dusk-balance"
>
  <header
    class="dusk-balance__header"
  >
    <h2
      class="sr-only"
    >
      Your Balance:
    </h2>
  </header>
   
  <p
    class="dusk-balance__dusk"
  >
    <span>
      0.003000000
    </span>
     
    <span>
      DUSK
    </span>
  </p>
   
  <p
    class="dusk-balance__fiat dusk-balance__fiat--visible"
  >
    <span>
      $0.03
    </span>
  </p>
   
  <div
    class="dusk-card dusk-card--gap-default dusk-balance__usage-details"
  >
     
    <div
      class="dusk-card__body-container"
    >
      <div
        class="dusk-balance__account"
      >
        <span
          class="dusk-balance__percentage"
        >
          <svg
            class="dusk-icon dusk-icon--size--default"
            data-tooltip-id="main-tooltip"
            data-tooltip-text="Shielded"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M12,1L3,5V11C3,16.55 6.84,21.74 12,23C17.16,21.74 21,16.55 21,11V5L12,1M12,7C13.4,7 14.8,8.1 14.8,9.5V11C15.4,11 16,11.6 16,12.3V15.8C16,16.4 15.4,17 14.7,17H9.2C8.6,17 8,16.4 8,15.7V12.2C8,11.6 8.6,11 9.2,11V9.5C9.2,8.1 10.6,7 12,7M12,8.2C11.2,8.2 10.5,8.7 10.5,9.5V11H13.5V9.5C13.5,8.7 12.8,8.2 12,8.2Z"
            />
          </svg>
          
          33.33
          %
        </span>
         
        <span
          class="dusk-balance__value"
        >
          0.001000000
          <svg
            class="dusk-icon dusk-icon--size--default"
            data-tooltip-id="main-tooltip"
            data-tooltip-text="DUSK"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M12.335 0.005a12.186 12.186 0 0 0 -2.58 0.204C4.2 1.259 0 6.14 0 11.999 0 17.859 4.203 22.74 9.759 23.79c0.726 0.138 1.476 0.21 2.241 0.21 6.789 0 12.261 -5.64 11.991 -12.49C23.741 5.249 18.599 0.173 12.335 0.005Zm0.207 21.58c-0.204 0.012 -0.336 -0.222 -0.21 -0.384 1.974 -2.541 3.153 -5.733 3.153 -9.202 0 -3.468 -1.176 -6.663 -3.153 -9.202 -0.126 -0.162 0.003 -0.396 0.207 -0.384 5.052 0.282 9.06 4.465 9.06 9.586 0 5.121 -4.005 9.304 -9.057 9.586Z"
            />
          </svg>
          
        </span>
      </div>
       
      <div
        class="dusk-balance__account"
      >
        <span
          class="dusk-balance__percentage"
        >
          <svg
            class="dusk-icon dusk-icon--size--default"
            data-tooltip-id="main-tooltip"
            data-tooltip-text="Public"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M21 11C21 16.5 17.2 21.7 12 23C6.8 21.7 3 16.5 3 11V5L12 1L21 5V11M12 21C15.8 20 19 15.5 19 11.2V6.3L12 3.2L5 6.3V11.2C5 15.5 8.2 20 12 21M14.8 11H10.5V8.5C10.5 7.7 11.2 7.2 12 7.2S13.5 7.7 13.5 8.5V9H14.8V8.5C14.8 7.1 13.4 6 12 6S9.2 7.1 9.2 8.5V11C8.6 11 8 11.6 8 12.2V15.7C8 16.4 8.6 17 9.2 17H14.7C15.4 17 16 16.4 16 15.8V12.3C16 11.6 15.4 11 14.8 11Z"
            />
          </svg>
          
          66.67
          %
        </span>
         
        <span
          class="dusk-balance__value"
        >
          0.002000000
          <svg
            class="dusk-icon dusk-icon--size--default"
            data-tooltip-id="main-tooltip"
            data-tooltip-text="DUSK"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M12.335 0.005a12.186 12.186 0 0 0 -2.58 0.204C4.2 1.259 0 6.14 0 11.999 0 17.859 4.203 22.74 9.759 23.79c0.726 0.138 1.476 0.21 2.241 0.21 6.789 0 12.261 -5.64 11.991 -12.49C23.741 5.249 18.599 0.173 12.335 0.005Zm0.207 21.58c-0.204 0.012 -0.336 -0.222 -0.21 -0.384 1.974 -2.541 3.153 -5.733 3.153 -9.202 0 -3.468 -1.176 -6.663 -3.153 -9.202 -0.126 -0.162 0.003 -0.396 0.207 -0.384 5.052 0.282 9.06 4.465 9.06 9.586 0 5.121 -4.005 9.304 -9.057 9.586Z"
            />
          </svg>
          
        </span>
      </div>
    </div>
     
  </div>
  
</article>
`;

exports[`Balance > should update the Balance component when the props change 2`] = `
<article
  class="dusk-balance"
>
  <header
    class="dusk-balance__header"
  >
    <h2
      class="sr-only"
    >
      Your Balance:
    </h2>
  </header>
   
  <p
    class="dusk-balance__dusk"
  >
    <span>
      0,003000000
    </span>
     
    <span>
      DUSK
    </span>
  </p>
   
  <p
    class="dusk-balance__fiat dusk-balance__fiat--visible"
  >
    <span>
      0,06 €
    </span>
  </p>
   
  <div
    class="dusk-card dusk-card--gap-default dusk-balance__usage-details"
  >
     
    <div
      class="dusk-card__body-container"
    >
      <div
        class="dusk-balance__account"
      >
        <span
          class="dusk-balance__percentage"
        >
          <svg
            class="dusk-icon dusk-icon--size--default"
            data-tooltip-id="main-tooltip"
            data-tooltip-text="Shielded"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M12,1L3,5V11C3,16.55 6.84,21.74 12,23C17.16,21.74 21,16.55 21,11V5L12,1M12,7C13.4,7 14.8,8.1 14.8,9.5V11C15.4,11 16,11.6 16,12.3V15.8C16,16.4 15.4,17 14.7,17H9.2C8.6,17 8,16.4 8,15.7V12.2C8,11.6 8.6,11 9.2,11V9.5C9.2,8.1 10.6,7 12,7M12,8.2C11.2,8.2 10.5,8.7 10.5,9.5V11H13.5V9.5C13.5,8.7 12.8,8.2 12,8.2Z"
            />
          </svg>
          
          16,67
          %
        </span>
         
        <span
          class="dusk-balance__value"
        >
          0,000500000
          <svg
            class="dusk-icon dusk-icon--size--default"
            data-tooltip-id="main-tooltip"
            data-tooltip-text="DUSK"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M12.335 0.005a12.186 12.186 0 0 0 -2.58 0.204C4.2 1.259 0 6.14 0 11.999 0 17.859 4.203 22.74 9.759 23.79c0.726 0.138 1.476 0.21 2.241 0.21 6.789 0 12.261 -5.64 11.991 -12.49C23.741 5.249 18.599 0.173 12.335 0.005Zm0.207 21.58c-0.204 0.012 -0.336 -0.222 -0.21 -0.384 1.974 -2.541 3.153 -5.733 3.153 -9.202 0 -3.468 -1.176 -6.663 -3.153 -9.202 -0.126 -0.162 0.003 -0.396 0.207 -0.384 5.052 0.282 9.06 4.465 9.06 9.586 0 5.121 -4.005 9.304 -9.057 9.586Z"
            />
          </svg>
          
        </span>
      </div>
       
      <div
        class="dusk-balance__account"
      >
        <span
          class="dusk-balance__percentage"
        >
          <svg
            class="dusk-icon dusk-icon--size--default"
            data-tooltip-id="main-tooltip"
            data-tooltip-text="Public"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M21 11C21 16.5 17.2 21.7 12 23C6.8 21.7 3 16.5 3 11V5L12 1L21 5V11M12 21C15.8 20 19 15.5 19 11.2V6.3L12 3.2L5 6.3V11.2C5 15.5 8.2 20 12 21M14.8 11H10.5V8.5C10.5 7.7 11.2 7.2 12 7.2S13.5 7.7 13.5 8.5V9H14.8V8.5C14.8 7.1 13.4 6 12 6S9.2 7.1 9.2 8.5V11C8.6 11 8 11.6 8 12.2V15.7C8 16.4 8.6 17 9.2 17H14.7C15.4 17 16 16.4 16 15.8V12.3C16 11.6 15.4 11 14.8 11Z"
            />
          </svg>
          
          83,33
          %
        </span>
         
        <span
          class="dusk-balance__value"
        >
          0,002500000
          <svg
            class="dusk-icon dusk-icon--size--default"
            data-tooltip-id="main-tooltip"
            data-tooltip-text="DUSK"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M12.335 0.005a12.186 12.186 0 0 0 -2.58 0.204C4.2 1.259 0 6.14 0 11.999 0 17.859 4.203 22.74 9.759 23.79c0.726 0.138 1.476 0.21 2.241 0.21 6.789 0 12.261 -5.64 11.991 -12.49C23.741 5.249 18.599 0.173 12.335 0.005Zm0.207 21.58c-0.204 0.012 -0.336 -0.222 -0.21 -0.384 1.974 -2.541 3.153 -5.733 3.153 -9.202 0 -3.468 -1.176 -6.663 -3.153 -9.202 -0.126 -0.162 0.003 -0.396 0.207 -0.384 5.052 0.282 9.06 4.465 9.06 9.586 0 5.121 -4.005 9.304 -9.057 9.586Z"
            />
          </svg>
          
        </span>
      </div>
    </div>
     
  </div>
  
</article>
`;
